### PR TITLE
[MERGE] mail: remove composer onchange and clean its usage

### PR DESCRIPTION
**Pull request by tde-banana-odoo**
PURPOSE

Purpose of this task is to remove the ``_onchange_template_id`` method on
composer model and replace it by computed fields. Indeed onchange have to
be manually called. It is easy to forget calling it in various processes and
have different results depending on composer usage.

SPECIFICATIONS

Remove ``_onchange_template_id`` and its wrapper. Move impacted fields
into editable computed stored fields.

While doing this some extra cleaning is required. Notably on template model
values generation is now better split (main, recipients, attachments). That
way it can easily be called from computed fields.

Tests are added and updated accordingly. Almost all tests behave like before
this task.

LINKS

Task-2088884 (Mail: Composer Onchange to Editable Computed Stored)
Task-2710804 (Mail: Clean MailThread API)
Task-2684479 (Mail: Better report errors when sending emails)
Prepares Task-36879 (Mail: Support MultiCompany Aliases)

### CODE REVIEW

**Comment by Feyensv** [comment]
Pls ping me when I can fully review it :eyes:

**Comment by tde-banana-odoo** [comment]
@Feyensv Sure :)

File: addons/mail/wizard/mail_compose_message.py [hunk 1]
```diff
@@ -161,21 +271,39 @@ def _inverse_reply_to_mode(self):
         for composer in self:
             composer.reply_to_force_new = composer.reply_to_mode == 'new'
 
+    @api.depends('composition_mode', 'model', 'parent_id', 'res_id', 'template_id')
+    def _compute_partner_ids(self):
+        """ Recipients computation.
+
+        """
```
Comment (std-odoo) [inline addons/mail/wizard/mail_compose_message.py:278]: not useful docstring imo

File: addons/mail/models/mail_render_mixin.py [hunk 2]
```diff
@@ -435,12 +441,15 @@ def _render_template(self, template_src, model, res_ids, engine='inline_template
...
-        :param boolean post_process: perform a post processing on rendered result
-          (notably html links management). See``_render_template_postprocess``;
+        :param dict options: options for rendering; may contain
+            boolean post_process: perform a post processing on rendered result
+            (notably html links management). See``_render_template_postprocess``);
 
         :return dict: {res_id: string of rendered template based on record}
         """
+        if options is None:
+            options = {}
```
Comment (std-odoo) [inline addons/mail/models/mail_render_mixin.py:451]:
```suggestion
        options = options or {}
```

File: addons/mail/models/mail_template.py [hunk 3]
```diff
@@ -135,113 +137,173 @@ def create_action(self):
...
-        Emails (email_to, email_cc) can be transformed into partners if requested
-        in the context. """
+    def _generate_template_recipients(self, res_ids, render_fields, partners_only=False, render_results=None):
+        """ Render recipients of the template 'self', returning values for records
+        given by 'res_ids'. Default values can be generated instead of the template
+        values if requested by template. Emails (email_to, email_cc) can be
+        transformed into partners if requested. """
         self.ensure_one()
+        if render_results is None:
+            render_results = dict()
```
Comment (std-odoo) [inline addons/mail/models/mail_template.py:147]:
```suggestion
        render_results = render_results or {}
```

File: addons/mail/models/mail_template.py [hunk 4]
```diff
@@ -135,113 +137,173 @@ def create_action(self):
...
+        given by 'res_ids'.
+
+        :param list render_fields: list of fields to render. Should contain
+          either attachments (reports) or attachment_ids (template attachments);
+        :param render_results: if given fill it instead of returning a new
+          dict;
+        """
+        self.ensure_one()
+        if render_results is None:
+            render_results = dict()
```
Comment (std-odoo) [inline addons/mail/models/mail_template.py:211]:
```suggestion
        render_results = render_results or {}
```

File: odoo/addons/base/tests/common.py [hunk 5]
```diff
@@ -251,15 +254,23 @@ def login(self, user, password):
...
         IrMailServer = self.env['ir.mail_server']
-        connect = IrMailServer.connect
-        find_mail_server = IrMailServer._find_mail_server
+        connect_origin = IrMailServer.connect
+        find_mail_server_origin = IrMailServer._find_mail_server
+
+        def config_get(*args, **kwargs):
+            if args[0] == 'smtp_server':
+                return self.config_smtp_server
+            return get_origin(*args, **kwargs)
```
Comment (std-odoo) [inline odoo/addons/base/tests/common.py:265]: isn't it better to return `None` ? so we simulate the fact that no config is configured in the odoo-bin arguments, and we don't need the argument `config_smtp_server`

File: odoo/addons/base/tests/test_ir_mail_server.py [hunk 6]
```diff
@@ -336,7 +337,7 @@ def test_mail_server_binary_arguments_domain(self):
         self.assertFalse(self.env['ir.mail_server'].search([]))
 
         # Use an email in the domain of the "from_filter"
-        with self.mock_smtplib_connection():
+        with self.mock_smtplib_connection(config_smtp_server='localhost'):
```
Comment (std-odoo) [inline odoo/addons/base/tests/test_ir_mail_server.py:340]: why is this required ? (it's weird that it try to read the odoo-bin SMTP config, because you created mail server in `_init_mail_servers`)

(same for all other example below)

File: odoo/addons/base/tests/common.py [hunk 7]
```diff
@@ -251,15 +254,23 @@ def login(self, user, password):
...
 
-        with patch.object(type(IrMailServer), '_is_test_mode', lambda self: False), \
-             patch('smtplib.SMTP_SSL', side_effect=lambda *args, **kwargs: self.testing_smtp_session), \
+        with patch('smtplib.SMTP_SSL', side_effect=lambda *args, **kwargs: self.testing_smtp_session), \
              patch('smtplib.SMTP', side_effect=lambda *args, **kwargs: self.testing_smtp_session), \
-             patch.object(type(IrMailServer), 'connect', side_effect=connect) as connect_mocked, \
-             patch.object(type(IrMailServer), '_find_mail_server', side_effect=find_mail_server) as find_mail_server_mocked:
+             patch.object(config, 'get', side_effect=config_get) as config_mocked, \
```
Comment (std-odoo) [inline odoo/addons/base/tests/common.py:269]: not sure this mock is necessary
```suggestion
```

File: addons/test_mail/tests/test_message_post.py [hunk 8]
```diff
@@ -211,115 +241,308 @@ def test_notify_recipients_internals(self):
...
+    @users('employee')
+    def test_notify_batch(self):
+        """ Test notify in batch. Currently not supported actually. """
+        test_records, _ = self._create_records_for_batch('mail.test.simple', 10)
+
+        with self.assertRaises(ValueError):
+            test_records.message_notify(
+                body='<p>Nice notification content</p>',
+                subject='Notify Subject',
+                partner_ids=self.partner_employee_2.ids
```
Comment (std-odoo) [inline addons/test_mail/tests/test_message_post.py:297]:
```suggestion
                partner_ids=self.partner_employee_2.ids,
```

File: addons/test_mail/tests/test_message_post.py [hunk 9]
```diff
@@ -17,12 +18,30 @@
...
+        # users (and not admin / user)
+        cls.user_employee_2 = mail_new_test_user(
+            cls.env, login='employee2',
+            groups='base.group_user',
+            company_id=cls.company_admin.id,
+            # email='"Eglantine Employee2" <eglantine@example.com>',
+            email='eglantine@example.com',
+            name='Eglantine Employee2',
+            notification_type='email',
+            signature='--\nEglantine'
```
Comment (std-odoo) [inline addons/test_mail/tests/test_message_post.py:41]: such a review
```suggestion
            signature='--\nEglantine',
```

File: addons/test_mail/tests/test_message_post.py [hunk 10]
```diff
@@ -211,115 +241,308 @@ def test_notify_recipients_internals(self):
...
+             'res_id': test_record.id,
+             'subtype_id': self.env.ref('mail.mt_note'),
+            }
+        )
+        self.assertNotIn(new_notification, self.test_record.message_ids)
+
+        admin_mails = [x for x in self._mails if self.partner_admin.name in x.get('email_to')[0]]
+        self.assertEqual(len(admin_mails), 1, 'There should be exactly one email sent to admin')
+        admin_mail = admin_mails[0].get('body')
+        admin_access_link = admin_mail[admin_mail.index('model='):admin_mail.index('/>') - 1] if 'model=' in admin_mail else None
```
Comment (std-odoo) [inline addons/test_mail/tests/test_message_post.py:277]: a regex is a bit more explicit imo, as you prefer
```suggestion
        admin_access_link = re.search(r'model=[\w]+', admin_mail).group()
```

File: addons/test_mail/tests/test_message_management.py [hunk 11]
```diff
@@ -36,6 +36,7 @@ def test_mail_resend_workflow(self):
                 [{'partner': partner, 'type': 'email', 'status': 'exception'} for partner in self.partners],
                 message_info={'message_type': 'notification'},
                 sim_error='connect_failure'):
+            self.connect_mocked.side_effect = Exception("Some exception")
```
Comment (std-odoo) [inline addons/test_mail/tests/test_message_management.py:39]: shouldn't you do:
```suggestion
            def _connect(self):
                raise Exception("Some exception")
            self.connect_mocked.side_effect = _connect
```

I think it will indeed fail, but for the wrong reason

**Review by std-odoo — COMMENTED** [review]
Early validation 🦆

(I'll have a deeper look when runbot is happy)

File: addons/mail/wizard/mail_compose_message.py [hunk 12]
```diff
@@ -644,24 +695,35 @@ def render_message(self, res_ids):
...
 
-        return multi_mode and values or values[res_ids[0]]
+        return values
+
+    # ------------------------------------------------------------
+    # TOOLS
+    # ------------------------------------------------------------
+
+    def _set_value_from_template(self, template_fname, composer_fname=False, force_void=False):
+        composer_fname = composer_fname if composer_fname else template_fname
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/wizard/mail_compose_message.py:721]: My linter does not agree with you on this one :) "always use ternary form". I tend to agree with him, I prefer if-else construct.

File: addons/mail/models/mail_render_mixin.py (hunk 2 above)
Comment (tde-banana-odoo) [reply addons/mail/models/mail_render_mixin.py:451]: Actually doing that you create a dict when you give a void dict which is useless. I think this is more an old habit of mine, either you give None (default) and I instantiate a new dict, either I use the parameter given. If you give False that's your fault.

File: odoo/addons/base/tests/common.py (hunk 5 above)
Comment (tde-banana-odoo) [reply odoo/addons/base/tests/common.py:265]: It could have a side effect on other stuff... there are a lot of "default configuration" elements, I would keep the mock as small as possible.

File: odoo/addons/base/tests/test_ir_mail_server.py (hunk 6 above)
Comment (tde-banana-odoo) [reply odoo/addons/base/tests/test_ir_mail_server.py:340]: Because here you test when not having any mail servers. If I don't set the smtp, all sub tests crash (no mail servers, no from, cracboum). I just update tests you wrote so maybe that's not what was intended to be tested ?

Comment (std-odoo) [reply odoo/addons/base/tests/test_ir_mail_server.py:340]: ah yes indeed, can't you do something like this ?
```python
@patch.dict("odoo.tools.config.options", {"from_filter": "test.com", "smtp_server": "example.com"})
```

File: addons/mail/wizard/mail_compose_message.py [hunk 13]
```diff
@@ -144,13 +129,138 @@ def default_get(self, fields):
...
+    @api.depends('template_id', 'composition_mode', 'parent_id', 'record_name')
+    def _compute_subject(self):
+        for composer in self:
+            if composer.template_id.subject:
+                composer._set_value_from_template('subject')
+            elif not composer.template_id or not composer.subject:
+                subject = composer.parent_id.subject if composer.parent_id else False
+                if not subject:
+                    subject = composer.record_name if composer.record_name else False
+                if subject:
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/wizard/mail_compose_message.py:152]: That's not an else, but we add the "Re" prefix only if something has been found. I changed the name of the intermediate variable to be clearer.

File: addons/test_mail/tests/test_message_post.py (hunk 10 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_message_post.py:277]: Actually due to the res_mode=..&res_id=.. I have to make several re or a complex one. I prefer plain stupid indexes :D .

**Review by tde-banana-odoo — COMMENTED** [review]
Moving slowly forward in first comments

File: odoo/addons/base/tests/test_ir_mail_server.py (hunk 6 above)
Comment (tde-banana-odoo) [reply odoo/addons/base/tests/test_ir_mail_server.py:340]: Indeed :)

**Comment by tde-banana-odoo** [comment]
@std-odoo Note that tests have been moved in https://github.com/odoo/odoo/pull/86393 for pre-merge (and keep only business code change here).

**Comment by tde-banana-odoo** [comment]
Work done (notably) at https://github.com/odoo/odoo/pull/99482 and https://github.com/odoo/odoo/pull/107356 .
